pkg/controller/rollout/trafficrouting/ingress/ingress.go

```diff
@@ -0,0 +1,298 @@
+	canaryIngress.Annotations, err = r.executeLuaForCanary(canaryIngress.Annotations, utilpointer.Int32(0), nil)
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) execute lua failed: %s", r.conf.RolloutNs, r.conf.RolloutName, err.Error())
+		return err
+	}
+	if err = r.Create(ctx, canaryIngress); err != nil {
+		klog.Errorf("rollout(%s/%s) create canary ingress failed: %s", r.conf.RolloutNs, r.conf.RolloutName, err.Error())
+		return err
+	}
+	klog.Infof("rollout(%s/%s) create canary ingress(%s) success", r.conf.RolloutNs, r.conf.RolloutName, util.DumpJSON(canaryIngress))
+	return nil
+}
+
+func (r *ingressController) EnsureRoutes(ctx context.Context, weight *int32, matches []rolloutv1alpha1.HttpRouteMatch) (bool, error) {
+	canaryIngress := &netv1.Ingress{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: r.conf.RolloutNs, Name: defaultCanaryIngressName(r.conf.TrafficConf.Name)}, canaryIngress)
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) get canary ingress failed: %s", r.conf.RolloutNs, r.conf.RolloutName, err.Error())
+		return false, err
+	}
+	newAnnotations, err := r.executeLuaForCanary(canaryIngress.Annotations, weight, matches)
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) execute lua failed: %s", r.conf.RolloutNs, r.conf.RolloutName, err.Error())
+		return false, err
+	}
+	if reflect.DeepEqual(canaryIngress.Annotations, newAnnotations) {
+		return true, nil
+	}
+	byte1, _ := json.Marshal(metav1.ObjectMeta{Annotations: canaryIngress.Annotations})
+	byte2, _ := json.Marshal(metav1.ObjectMeta{Annotations: newAnnotations})
+	patch, err := jsonpatch.CreateMergePatch(byte1, byte2)
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) create merge patch failed: %s", r.conf.RolloutNs, r.conf.RolloutName, err.Error())
+		return false, err
+	}
+	body := fmt.Sprintf(`{"metadata":%s}`, string(patch))
+	if err = r.Patch(ctx, canaryIngress, client.RawPatch(types.MergePatchType, []byte(body))); err != nil {
+		klog.Errorf("rollout(%s/%s) set canary ingress(%s) failed: %s", r.conf.RolloutNs, r.conf.RolloutName, canaryIngress.Name, err.Error())
+		return false, err
+	}
+	klog.Infof("rollout(%s/%s) set canary ingress(%s) annotations(%s) success", r.conf.RolloutNs, r.conf.RolloutName, canaryIngress.Name, util.DumpJSON(newAnnotations))
+	return false, nil
+}
+
+func (r *ingressController) Finalise(ctx context.Context) (bool, error) {
+	canaryIngress := &netv1.Ingress{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: r.conf.RolloutNs, Name: r.canaryIngressName}, canaryIngress)
+	if err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("rollout(%s/%s) get canary ingress(%s) failed: %s", r.conf.RolloutNs, r.conf.RolloutName, r.canaryIngressName, err.Error())
+		return false, err
+	}
+	if errors.IsNotFound(err) {
+		return true, nil
+	} else if !canaryIngress.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+	// First, set canary route 0 weight.
+	newAnnotations, err := r.executeLuaForCanary(canaryIngress.Annotations, utilpointer.Int32(0), nil)
```

Comment:
plz make sure that we use zero weight or -1 weight

Reply:
Here is setting weight=0, because I want to set zero traffic to canary service.